File: src/gemmini/gemmini_sim_object.cc
#include "gemmini/gemmini_sim_object.hh"
#include <iostream>
#include "base/trace.hh"
#include "debug/GemminiExampleFlag.hh"
#include "sim/eventq.hh"

namespace gem5
{
    GemminiSimObject::GemminiSimObject(const GemminiSimObjectParams& params): 
        SimObject(params),
        mode(0),
        relu(0),
        A_transpose(0),
        B_transpose(0),
        A_stride(1),     // 默认stride通常设为1
        scale(1.0f),     // 默认不缩放
        shift(0)         // 默认不shift
        {
            printf("-----------Gemmini initialized------------\n");
        }

    void
    GemminiSimObject::config_ex(uint64_t rs1, uint64_t rs2)
    {
        parseConfigEx(rs1, rs2);
        printConfig();
    }
    
    void 
    GemminiSimObject::parseConfigEx(uint64_t rs1, uint64_t rs2){
        // ===== 1. 解析 rs1 =====
        mode        = (rs1 >> 2) & 0x1;
        relu        = (rs1 >> 3) & 0x1;
        A_transpose = (rs1 >> 8) & 0x1;
        B_transpose = (rs1 >> 9) & 0x1;
        A_stride = (rs1 >> 16) & 0xFFFF;
        // ===== 2. 解析 scale（float32）=====
        uint32_t scale_bits = (rs1 >> 32) & 0xFFFFFFFF;
        std::memcpy(&scale, &scale_bits, sizeof(float));
        // ===== 3. 解析 rs2 =====
        shift = rs2 & 0xFFFFFFFF;
    }

    void 
    GemminiSimObject::printConfig(){
        DPRINTF(GemminiExampleFlag, "=== config_ex ===\n");
        DPRINTF(GemminiExampleFlag, "mode            = %d (%s-stationary)\n",
                mode, mode ? "weight" : "output");
        DPRINTF(GemminiExampleFlag, "relu            = %d\n", relu);
        DPRINTF(GemminiExampleFlag, "A_transpose     = %d\n", A_transpose);
        DPRINTF(GemminiExampleFlag, "B_transpose     = %d\n", B_transpose);
        DPRINTF(GemminiExampleFlag, "A_stride        = %d\n", A_stride);
        DPRINTF(GemminiExampleFlag, "scale           = %f\n", scale);
        DPRINTF(GemminiExampleFlag, "shift           = %u\n", shift);
    }

} // namespace gem5

File: src/gemmini/GemminiSimObject.py
# 导入gem5所有模块的基类：SimObject（所有自定义硬件模块必须继承它）
from m5.objects.SimObject import SimObject
# 导入gem5的参数定义工具（Param是核心，用来给模块加配置）
from m5.params import *

# 自定义模块：GemminiSimObject
class GemminiSimObject(SimObject):
    # 1. gem5内部唯一识别名（必须和C++类名一致）
    type = "GemminiSimObject"
    # 2. 绑定：这个Python类对应的C++头文件路径
    cxx_header = "gemmini/gemmini_sim_object.hh"
    # 3. 绑定：这个Python类对应的C++类全名（带命名空间）→ C++gem5::HelloSimObject
    cxx_class = "gem5::GemminiSimObject"
    